src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTaskBase.cs

```diff
@@ -36,49 +36,106 @@
         {
             _serverStore = serverStore;
-            _restoreConfiguration = restoreConfiguration;
+            RestoreFromConfiguration = restoreFromConfiguration;
             _nodeTag = nodeTag;
             _operationCancelToken = operationCancelToken;
 
-            ValidateArguments(out _restoringToDefaultDataDirectory);
+            if (string.IsNullOrWhiteSpace(RestoreFromConfiguration.DatabaseName))
+                throw new ArgumentException("Database name can't be null or empty");
+
+            if (ResourceNameValidator.IsValidResourceName(RestoreFromConfiguration.DatabaseName, _serverStore.Configuration.Core.DataDirectory.FullPath, out string errorMessage) == false)
+                throw new BadRequestException(errorMessage);
+
+            _serverStore.EnsureNotPassive();
+
+            using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                if (_serverStore.Cluster.ReadDatabase(context, RestoreFromConfiguration.DatabaseName) != null)
+                    throw new ArgumentException($"Cannot restore data to an existing database named {RestoreFromConfiguration.DatabaseName}");
+
+                var clusterTopology = _serverStore.GetClusterTopology(context);
+
+                if (string.IsNullOrWhiteSpace(RestoreFromConfiguration.EncryptionKey) == false)
+                {
+                    var key = Convert.FromBase64String(RestoreFromConfiguration.EncryptionKey);
+                    if (key.Length != 256 / 8)
+                        throw new InvalidOperationException($"The size of the key must be 256 bits, but was {key.Length * 8} bits.");
+
+                    var isEncrypted = string.IsNullOrWhiteSpace(RestoreFromConfiguration.EncryptionKey) == false;
+                    if (isEncrypted && AdminDatabasesHandler.NotUsingHttps(clusterTopology.GetUrlFromTag(_serverStore.NodeTag)))
+                        throw new InvalidOperationException("Cannot restore an encrypted database to a node which doesn't support SSL!");
+                }
+            }
+
+            var hasRestoreDataDirectory = string.IsNullOrWhiteSpace(RestoreFromConfiguration.DataDirectory) == false;
+            if (hasRestoreDataDirectory &&
+                HasFilesOrDirectories(RestoreFromConfiguration.DataDirectory))
+                throw new ArgumentException("New data directory must be empty of any files or folders, " +
+                                            $"path: {RestoreFromConfiguration.DataDirectory}");
+
+            if (hasRestoreDataDirectory == false)
+                RestoreFromConfiguration.DataDirectory = GetDataDirectory();
+
+            _restoringToDefaultDataDirectory = IsDefaultDataDirectory(RestoreFromConfiguration.DataDirectory, RestoreFromConfiguration.DatabaseName);
+
+            _hasEncryptionKey = string.IsNullOrWhiteSpace(RestoreFromConfiguration.EncryptionKey) == false;
```

Comment:
I think this is the 3rd time we are checking that. Can we move that earlier and use the `_hasEncryptionKey` ?

Reply:
Yes. moved up and removed duplicate.